### ISSUE TITLE
join_asof

### ISSUE THREAD

**Issue by chrisflesher** [more-itertools/more-itertools#927]
### Description
Similar to polars [join_asof](https://docs.pola.rs/api/python/stable/reference/dataframe/api/polars.DataFrame.join_asof.html) but for streaming data.

### References
A common use case would be to synchronize streams by timestamp.

### Examples
Here is an example that is similar to polars `join_asof` with "backward" strategy, the summary is:
1. create asynchronous iterables `data` and `status`
2. synchronize these using `join_asof`
3. remove data with bad status using `filter_map`

```
import more_itertools
import typing

def join_asof(left: typing.Iterable,
              right: typing.Iterable,
              ) -> typing.Generator[typing.Tuple, None, None]:
    """Synchronize two iterables."""
    right_peekable = more_itertools.peekable(right)
    right_item = next(right_peekable)
    for left_item in left:
        if left_item < right_item:
            yield (left_item, None)
        else:
            while True:
                try:
                    right_peek = right_peekable.peek()
                except StopIteration:
                    break
                if left_item < right_peek:
                    break
                right_item = next(right_peekable)
            yield (left_item, right_item)


class Status(typing.NamedTuple):

    time_ns: int
    is_good: bool


class Data(typing.NamedTuple):

    time_ns: int
    measurement: float


statuses = [
    Status(time_ns=200, is_good=True),
    Status(time_ns=300, is_good=False),
]
data = [
    Data(time_ns=100, measurement=1.0),
    Data(time_ns=200, measurement=2.0),
    Data(time_ns=250, measurement=2.5),
    Data(time_ns=300, measurement=3.0),
    Data(time_ns=400, measurement=4.0),
]
data_and_statuses = list(join_asof(iter(data), iter(statuses)))
good_data = list(more_itertools.filter_map(lambda x: x[0] if x[1] is not None and x[1].is_good else None, data_and_statuses))
```

**Comment by rhettinger** [comment]
A few thoughts:

* This seems like a reasonable addition because there is no easy way for a user to do this with existing tools.
* The implementation works with `NamedTuple` instances that both have a timestamp in position zero and where the remaining fields are comparable.
* So the tool wouldn't work if `is_good` was a string. That couldn't be compared to the float in `measurement`.
* Also the tool won't work with dataclasses which are not cross-class comparable.
* A reasonable fix would be to have two key-functions (one for the left and one for the right). These could be an `itemgetter` or `attrgetter`.  That would let the fields be in any order and eliminate type restrictions on the other fields.
* Besides `peekable`, there are a number of ways to implement this perhaps using `spy`, `heapq.merge` or just calling `next` directly.
* If `right` is empty, it may be better to `return` than to `raise`.

**Comment by bbayles** [comment]
Can we think of a practical example of needing two different key functions? `heapq.merge` gets by with a single one.

Here's an attempt at a `merge`-based implementation:

```python
def synchronize(left, right, key=lambda x: x):
    left_data = ((0, x) for x in left)
    right_data = ((1, x) for x in right)
    seen_right = False
    last_right = None
    for source, data in merge(right_data, left_data, key=lambda x: key(x[1])):
        if source:
            seen_right = True
            last_right = data
        else:
            yield (data, last_right)
```

**Comment by rhettinger** [comment]
> Can we think of a practical example of needing two different key functions? heapq.merge gets by with a single one.

The polars API allows a single key function called `on` or two distinct key functions `left_on` and `right_on`. The reason is that the two input streams typically would have different types and may have differing field names. 

     join_asof(Iterator[S], Iterator[T]) -> Iterator[Tuple[S, T | None]]

In contrast, the `heapq.merge` input streams would typically be of the same type so one key function would always suffice:

    `merge(Iterator[T], Iterator[T]) -> Iterator[T]`

Here's a practical example:

```python
@dataclass
class Data(typing.NamedTuple):

    nanoseconds: int
    measurement: float

@dataclass
class Status:
    is_good: str
    ns: int

json_asoc(data, statuses, left_on=attrgetter('nanoseconds'), right_on=attrgetter('ns'))
```

That said, you could require that the same key-function apply to both inputs. It is a matter of how flexible you want to be.

> Here's an attempt at a merge-based implementation:

The `source, data` tuple is clever. I like it :-)

Instead of an identify function, the default key function should be `itemgetter(0)` so it will work tuple inputs where field-0 is the merge key and no assumptions are made about the other fields.

**Comment by bbayles** [comment]
I suppose the analog would be database tables that can be joined even though they have different column names.

> Instead of an identify function, the default key function should be itemgetter(0) so it will work tuple inputs where field-0 is the merge key and no assumptions are made about the other fields.

Wouldn't the identity function provide this? Tuple sorting would look at the 0 element first, right?

**Comment by rhettinger** [comment]
> Wouldn't the identity function provide this? Tuple sorting would look at the 0 element first, right?

Yes. However, if the values at position 0 are equal, the tuple comparison would look at the next field where the comparison isn't useful, may not make any sense, and could fail:

   left = [(1746217329, 'left')]
   right = [(1746217329, 10)]

**Comment by bbayles** [comment]
I've need things that rhyme with this before, but not quite exactly this. I think I'm -1 for inclusion based on that, and the fact that there don't seem to be any search results showing that  the existing Polars `join_asof` is an answer to anyone's questions on StackOverflow.

As always, I'm open to changing my mind provided examples of people trying to solve this.
